src/compute-client/src/service.rs

```diff
@@ -85,16 +85,53 @@ where
 
 /// Maintained state for partitioned compute clients.
 ///
-/// This helper type unifies the responses of multiple partitioned
-/// workers in order to present as a single worker.
+/// This helper type unifies the responses of multiple partitioned workers in order to present as a
+/// single worker:
+///
+///   * It emits `FrontierUppers` responses reporting the minimum/meet of frontiers reported by the
+///     individual workers.
+///   * It emits `PeekResponse`s and `SubscribeResponse`s reporting the union of the responses
+///     received from the workers.
+///
+/// In the compute communication stack, this client is instantiated several times:
+///
+///   * One instance on the controller side, dispatching between cluster processes.
+///   * One instance in each cluster process, dispatching between timely worker threads.
+///
+/// Note that because compute commands, except `CreateTimely`, are only sent to the first process,
+/// the cluster-side instances of `PartitionedComputeState` are not guaranteed to see all compute
+/// commands. Or more specifically: The instance running inside process 0 sees all commands,
+/// whereas the instances running inside the other processes only see `CreateTimely`. The
+/// `PartitionedComputeState` implementation must be able to cope with this limited visiblity. It
+/// does so by performing most of its state management based on observed compute responses rather
+/// than commands.
 #[derive(Debug)]
 pub struct PartitionedComputeState<T> {
     /// Number of partitions the state machine represents.
     parts: usize,
     /// Upper frontiers for indexes and sinks, both unioned across all partitions and from each
     /// individual partition.
+    ///
+    /// Frontier tracking for a collection is initialized when the first `FrontierUppers` response
+    /// for that collection is received. Frontier tracking is ceased when all shards have reported
+    /// advancement to the empty frontier.
+    ///
+    /// The compute protocol requires that shards always emit a `FrontierUppers` response reporting
+    /// the empty frontier when a collection is dropped. It further requires that no further
+    /// `FrontierUppers` responses are emitted for a collection after the empty frontier was
```

Comment:
What if for some reason we report the `FrontierUppers(empty)` more than once? Will this break something?

Reply:
If a replica emits `FrontierUppers(empty)` more than once for the same collection, we can at least expect state to leak, be cause point b) mentioned two lines below is not ensured anymore. The `PartitionedState` might already have dropped its frontier tracking state for the collection, but seeing a new `FrontierUppers` response will cause it to initialize new tracking state that then never gets cleaned up.

I think the controller would be similarly confused, with the difference that it sees all the commands and so realizes that it never has seen a `CreateDataflows` that corresponds to the stray `FrontierUppers`. So it will not initialize tracking state but instead log an error and ignore the response otherwise.

Reply:
> The PartitionedState might already have dropped its frontier tracking state for the collection, but seeing a new FrontierUppers response will cause it to initialize new tracking state that then never gets cleaned up.

My point is that maybe `FrontierUppers(empty)` does not warrant the creation of a tracking state... because if you know that the frontier is already empty and can only advance you might be able to reason that there won't be anything there to track... but I might be completely off with this, so feel free to ignore me.

Reply:
The issue is that we still might need to emit a `FrontierUppers(empty)` response to our upstream. It might be that we created a dataflow that immediately advanced to the empty frontier. The controller needs to know about that advancement so it can release read holds and cleanup state for that dataflow. So the partitioned client needs to initialize tracking and wait until it sees the `FrontierUppers(empty)` response from all shards. Only then can it forward that upstream (e.g. to the controller).

Of course, when a shard sends `FrontierUppers(empty)` twice, there is no need to forward it upstream again, since that has already been done. But if the partitioned client already dropped its tracking state for the collection, it doesn't remember forwarding the response, so it won't be able to detect this case.